### PR TITLE
Add support for Mkdocs search

### LINKED ISSUE #1088: ReadTheDocs search and MkDocs
(Sorry for adding this as an issue @ericholscher, but we keep missing each other on IRC. If there is a better place, let me know.)

We added the `mkdocs json` command to MkDocs for RTD's to use in serverside search. Since then I have been [working on clientside search](https://github.com/tomchristie/mkdocs/pull/222) support. This means we generate a JSON index in a slightly different format to work with Tipue search and thus we will potentially have two different JSON output formats which is a bit of a pain.

The Tipue format looks like this: https://github.com/Tipue/Tipue-Search/blob/master/tipuesearch/tipuesearch_content.js . The key difference is that it is one file containing all pages and the key names are different but otherwise I think it contains pretty much the same data.

So, my question is, can RTD be adapted to use this format so we can drop the specific json command from MkDocs? If so, I'd be happy to make the change in the ReadTheDocs code.

### PR DESCRIPTION
**Pull request by stsewd**
Closes #1088
Related to #4588

Things to do:

- ~~Override the search js with ours to call ES~~
- ~~Cache the processed index file per project~~
- ~~Add tests~~

But currently is working if someone wants to test it :)

Screenshot_2020-04-20 Search manually Read the Docs

Screenshot_2020-04-22 Read the Docs MkDocs Test(1)

Screenshot_2020-04-22 Read the Docs MkDocs Test

### CODE REVIEW

**Comment by stsewd** [comment]
Some comparisons (letft RTD search, right MkDocs search).
Screenshot from 2020-04-22 17-24-21
Screenshot from 2020-04-22 17-23-15
Screenshot from 2020-04-22 17-22-16
Screenshot from 2020-04-22 17-21-33
Screenshot from 2020-04-22 17-20-58
Screenshot from 2020-04-22 17-20-40
Screenshot from 2020-04-22 17-19-35

**Comment by stsewd** [comment]
To make the search available to all projects, we need to trigger a build or crawl all mkdocs projects (and versions) on storage and copy it's `search/search_index.json` file to the search media, after that we can trigger a re-index for those projects and all projects will start to use our search :)

**Comment by ericholscher** [comment]
@stsewd I wonder if it makes more sense to just read the `search_index.json` from the HTML output for mkdocs projects?

File: readthedocs/core/static-src/core/js/doc-embed/search.js [hunk 1]
```diff
@@ -288,9 +287,138 @@ function attach_elastic_search_query(data) {
...
+        initSearch();
+    });
+}
+
+
 function init() {
     var data = rtddata.get();
-    attach_elastic_search_query(data);
+    if (data.builder === 'mkdocs') {
+      attach_elastic_search_query_mkdocs(data);
```
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/core/static-src/core/js/doc-embed/search.js:418]: Sure, I can revert the js change and put it in another PR respecting the "skip search" flag.

File: readthedocs/projects/models.py [hunk 2]
```diff
@@ -1373,6 +1373,52 @@ def get_processed_json(self):
             'domain_data': {},
         }
 
+    def get_processed_json_mkdocs(self):
+        log.debug('Processing mkdocs index')
+        storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
+        storage_path = self.project.get_storage_path(
+            type_='json', version_slug=self.version.slug, include_file=False
```
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/projects/models.py:1380]: The json builder isn't being used right now, but yeah, we can probably just use the file from the media/html storage.

**Comment by stsewd** [comment]
Changed to index directly from the html artifacts in storage and reverted the js changes, this is ready for review.

**Pull request merged**